src/audio/kpb.c

```diff
@@ -1252,6 +1252,15 @@ static enum task_state kpb_draining_task(void *arg)
 
 		}
 	}
+
+	while (!sink->stream.free) {
+		/* This is to workaround HW problem of no DMA complete
+		 * IRQ. Because of that we have to be be sure HOST DMA
+		 * has some free space otherwise real time copy may be
+		 * stuck.
+		 */
+		comp_copy(sink->sink);
+	}
```

Comment:
This again looks like a loop instead of something based on timers and/or signaling. Not so good.

Reply:
Agree, I removed it and added better solution.